File: android/app/src/main/java/com/metallic/chiaki/common/DonationPhrasePicker.kt
// SPDX-License-Identifier: LicenseRef-AGPL-3.0-only-OpenSSL

package com.metallic.chiaki.common

import android.content.Context
import com.pylux.stream.R
import org.json.JSONObject

/**
 * Loads [R.raw.donation_prompt_phrases] copied at build time from [shared/donation_prompt_phrases.json]
 * (see `syncDonationPhrases` in `android/app/build.gradle`). Algorithm matches [shared/DONATION_PHRASE_PICKER.md].
 */
object DonationPhrasePicker
{
	/** Paywall opens 1..this many keep the default bullet list; after that, one rotating phrase replaces it. */
	const val PAYWALL_BULLET_SHOWS_BEFORE_PHRASES = 3

	private val defaultCategoryOrder = listOf("mild", "playful", "mean")

	@Volatile
	private var cachedFlat: List<String>? = null

	/** Clears cached phrases (e.g. after locale or asset change). */
	fun clearCache()
	{
		cachedFlat = null
	}

	/**
	 * 1-based call id: first phrase uses [callId] == 1, second == 2, … wraps over the flattened list.
	 */
	fun phraseForCallId(context: Context, callId: Int): String?
	{
		if (callId < 1) return null
		val flat = cachedFlat ?: loadFlat(context).also { cachedFlat = it }
		if (flat.isEmpty()) return null
		val i = floorMod(callId - 1, flat.size)
		return flat[i]
	}

	/**
	 * [showCount] = total paywall opens (1-based). First three opens → null (keep default bullets).
	 * From the 4th open, returns the rotating phrase (uses [showCount] so rotation advances each time).
	 */
	fun phraseForPaywallShowCount(context: Context, showCount: Int): String?
	{
		if (showCount <= PAYWALL_BULLET_SHOWS_BEFORE_PHRASES) return null
		val phraseCallId = showCount - PAYWALL_BULLET_SHOWS_BEFORE_PHRASES
		return phraseForCallId(context, phraseCallId)
	}

	private fun loadFlat(context: Context): List<String>
	{
		val text = context.resources.openRawResource(R.raw.donation_prompt_phrases).bufferedReader().use { it.readText() }
		val root = JSONObject(text)
		val order = root.optJSONArray("category_order")?.let { arr ->
			(0 until arr.length()).map { arr.getString(it) }
		} ?: defaultCategoryOrder
		val categories = root.optJSONObject("categories") ?: return emptyList()
		val out = ArrayList<String>()
		for (key in order)
		{
			val cat = categories.optJSONObject(key) ?: continue
			val phrases = cat.optJSONArray("phrases") ?: continue
			for (i in 0 until phrases.length())
				out.add(phrases.getString(i))
		}
		return out
	}

	private fun floorMod(a: Int, b: Int): Int
	{
		val m = a % b
		return if (m < 0) m + b else m
	}
}

File: ios/Pylux/Donation/DonationPhrasePicker.swift
// SPDX-License-Identifier: LicenseRef-AGPL-3.0-only-OpenSSL

import Foundation
import os.log

private let phraseLog = OSLog(subsystem: "com.pylux.stream", category: "DonationPhrasePicker")

enum DonationPhrasePicker {
    static let paywallBulletShowsBeforePhrases = 3

    private static let defaultCategoryOrder = ["mild", "playful", "mean"]

    private static var cachedFlat: [String]?

    static func clearCache() {
        cachedFlat = nil
    }

    /// 1-based call id: first phrase uses callId == 1, wraps over the flattened list.
    static func phrase(forCallId callId: Int) -> String? {
        guard callId >= 1 else { return nil }
        let flat = cachedFlat ?? loadFlat()
        guard !flat.isEmpty else { return nil }
        let i = floorMod(callId - 1, flat.count)
        return flat[i]
    }

    /// showCount = total paywall opens (1-based). First 3 -> nil (show bullets).
    /// From the 4th open, returns a rotating phrase.
    static func phrase(forPaywallShowCount showCount: Int) -> String? {
        guard showCount > paywallBulletShowsBeforePhrases else { return nil }
        let phraseCallId = showCount - paywallBulletShowsBeforePhrases
        return phrase(forCallId: phraseCallId)
    }

    // MARK: - Private

    private static func loadFlat() -> [String] {
        guard let url = Bundle.main.url(forResource: "donation_prompt_phrases", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let categories = root["categories"] as? [String: Any] else {
            os_log(.error, log: phraseLog, "Failed to load donation_prompt_phrases.json from bundle")
            return []
        }

        let order: [String]
        if let arr = root["category_order"] as? [String] {
            order = arr
        } else {
            order = defaultCategoryOrder
        }

        var out: [String] = []
        for key in order {
            guard let cat = categories[key] as? [String: Any],
                  let phrases = cat["phrases"] as? [String] else { continue }
            out.append(contentsOf: phrases)
        }
        cachedFlat = out
        return out
    }

    private static func floorMod(_ a: Int, _ b: Int) -> Int {
        let m = a % b
        return m < 0 ? m + b : m
    }
}
